File: router/routing.py
"""Event classification — the core of the service.

Everything else in this repo is plumbing around this module. It is deliberately
pure: no I/O, no clock, no network. Given a decoded Jetstream event it returns
the destination that event belongs to, or ``None`` if the event is malformed.
That purity is what makes the routing table exhaustively testable without
standing up a broker.

Jetstream event shapes (verified against the live firehose, not the docs)::

    # commit / create — carries a record
    {"did": "...", "time_us": 1785266126673819, "kind": "commit",
     "commit": {"rev": "...", "operation": "create", "rkey": "...",
                "collection": "app.bsky.feed.like",
                "record": {"subject": {"uri": "at://...", "cid": "..."}}}}

    # commit / delete — NOTE: no "record" key at all
    {"did": "...", "time_us": ..., "kind": "commit",
     "commit": {"rev": "...", "operation": "delete", "rkey": "...",
                "collection": "app.bsky.feed.post"}}

    # account (and the analogous "identity") — no "commit" key at all
    {"did": "...", "time_us": ..., "kind": "account",
     "account": {"active": false, "status": "deleted", "seq": ...}}

The absence of ``record`` on deletes is a real constraint on what the retraction
path can do; see ``workers/retraction.py`` and DESIGN.md.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from enum import StrEnum
from typing import Any

SUBJECT_PREFIX = "bsky"

# Jetstream's own vocabulary, spelled out so typos surface as NameErrors.
KIND_COMMIT = "commit"
OP_CREATE = "create"
OP_UPDATE = "update"
OP_DELETE = "delete"


class Destination(StrEnum):
    """One destination == one independently deployed, independently scaled worker.

    Adding a value here is a deliberate act: it implies a new Deployment, a new
    durable consumer, its own DLQ and its own scaling policy. Adding a *collection*
    to an existing destination is by contrast just a config change — see
    ``DEFAULT_COLLECTION_ROUTES`` and ``Settings.collection_routes``.
    """

    CONTENT = "content"
    ENGAGEMENT = "engagement"
    GRAPH = "graph"
    RETRACTION = "retraction"
    OTHER = "other"

    @property
    def subject(self) -> str:
        """NATS subject this destination's worker consumes from."""
        return f"{SUBJECT_PREFIX}.{self.value}"

    @property
    def dlq_subject(self) -> str:
        return f"{SUBJECT_PREFIX}.dlq.{self.value}"


# Which collection feeds which worker. Overridable at runtime from a ConfigMap
# (JSON in ROUTER_COLLECTION_ROUTES), so onboarding a new lexicon onto an
# existing path needs no image rebuild.
DEFAULT_COLLECTION_ROUTES: dict[str, Destination] = {
    "app.bsky.feed.post": Destination.CONTENT,
    "app.bsky.feed.like": Destination.ENGAGEMENT,
    "app.bsky.feed.repost": Destination.ENGAGEMENT,
    "app.bsky.graph.follow": Destination.GRAPH,
}


@dataclass(frozen=True, slots=True)
class Route:
    """The classification verdict for a single event."""

    destination: Destination
    kind: str
    collection: str | None
    operation: str | None
    did: str | None
    time_us: int | None
    #: Why this destination was chosen. Emitted as a metric label and a message
    #: header so a surprising routing decision is greppable in production
    #: rather than requiring a debugger.
    reason: str

    @property
    def subject(self) -> str:
        return self.destination.subject

    def headers(self) -> dict[str, str]:
        """Classification carried alongside the payload.

        Workers get the verdict for free instead of re-deriving it, and the
        headers are what an operator sees when inspecting a stuck message.
        """
        out = {
            "destination": self.destination.value,
            "kind": self.kind,
            "reason": self.reason,
        }
        if self.collection:
            out["collection"] = self.collection
        if self.operation:
            out["operation"] = self.operation
        if self.did:
            out["did"] = self.did
        if self.time_us is not None:
            out["time_us"] = str(self.time_us)
        return out


def classify(
    event: Any,
    collection_routes: Mapping[str, Destination] | None = None,
) -> Route | None:
    """Map a decoded Jetstream event to its destination.

    Returns ``None`` for anything structurally unusable. Callers count those
    separately: a malformed event is an upstream-schema signal, not just
    another unrouted event, and conflating the two hides schema drift.

    Precedence is deliberate and load-bearing:

    1. Non-commit kinds (``identity`` / ``account``) have no collection at all,
       so they can never match the collection table.
    2. **Deletes win over collection.** A deleted post goes to the retraction
       worker, not the content worker. The brief asks for retraction to be "a
       distinct path handled separately from creates", and a delete carries no
       record — so the content worker has nothing to match against anyway.
       Getting this backwards is the single easiest way to break the routing.
    3. Otherwise the collection table decides.
    4. Anything left over is OTHER, which is counted rather than dropped.
    """
    if not isinstance(event, Mapping):
        return None

    kind = event.get("kind")
    if not isinstance(kind, str) or not kind:
        return None

    did = event.get("did") if isinstance(event.get("did"), str) else None
    time_us = event.get("time_us") if isinstance(event.get("time_us"), int) else None

    # (1) identity / account — real events, no downstream work wired up yet.
    if kind != KIND_COMMIT:
        return Route(
            destination=Destination.OTHER,
            kind=kind,
            collection=None,
            operation=None,
            did=did,
            time_us=time_us,
            reason="non-commit-kind",
        )

    commit = event.get("commit")
    if not isinstance(commit, Mapping):
        return None

    collection = commit.get("collection")
    operation = commit.get("operation")
    if not isinstance(collection, str) or not collection:
        return None
    if not isinstance(operation, str) or not operation:
        return None

    # (2) Deletes take precedence over the collection table.
    if operation == OP_DELETE:
        return Route(
            destination=Destination.RETRACTION,
            kind=kind,
            collection=collection,
            operation=operation,
            did=did,
            time_us=time_us,
            reason="delete-precedence",
        )

    routes = DEFAULT_COLLECTION_ROUTES if collection_routes is None else collection_routes

    # (3) Collection table.
    destination = routes.get(collection)
    if destination is not None:
        return Route(
            destination=destination,
            kind=kind,
            collection=collection,
            operation=operation,
            did=did,
            time_us=time_us,
            reason="collection-map",
        )

    # (4) Known-shape event we have no path for. Counted, never silently dropped —
    # a rising OTHER rate is how we learn the network shipped a new lexicon.
    return Route(
        destination=Destination.OTHER,
        kind=kind,
        collection=collection,
        operation=operation,
        did=did,
        time_us=time_us,
        reason="unmapped-collection",
    )


def wanted_collections(
    collection_routes: Mapping[str, Destination] | None = None,
) -> list[str]:
    """Collections to request server-side via Jetstream's ``wantedCollections``.

    Derived from the routing table rather than configured separately, so the
    filter can never drift out of sync with what we actually know how to route.
    This is the cheapest backpressure in the system: events filtered here cost
    us no bandwidth, no parse and no broker write. Measured against the live
    firehose it is roughly an order of magnitude less traffic.

    Note this filter only constrains ``commit`` events — Jetstream still
    delivers ``identity`` and ``account`` regardless, which is why OTHER exists.
    """
    routes = DEFAULT_COLLECTION_ROUTES if collection_routes is None else collection_routes
    return sorted(routes)

File: internal/routing/routing.go
// Package routing maps an event to a route name.
//
// Everything here is a pure function: an event in, a route name out. No I/O, no
// state, no logging. That is what makes the routing table — the core of the
// exercise — exhaustively testable without standing up a socket, a channel or a
// worker.
//
// The table is configuration rather than code (a ConfigMap), so adding an event
// type to an existing route is a config change. Adding a genuinely new route is
// not, and should not be: a new route implies a buffer, a pool, a drop policy
// and its own metrics.
package routing

import (
	"fmt"
	"path"
	"strings"

	"github.com/shalevsharfsh/jetstream-router/internal/event"
)

// Match is one rule's predicate. An empty field matches anything.
//
// Collection accepts a glob ("app.bsky.feed.*"), because lexicon namespaces are
// hierarchical and a new sibling type usually belongs on the route its siblings
// already use.
type Match struct {
	Kind       string `json:"kind"`
	Collection string `json:"collection"`
	Operation  string `json:"operation"`
}

// Rule binds a predicate to a route.
type Rule struct {
	Match Match  `json:"match"`
	Route string `json:"route"`
}

// Table is the configured routing table.
type Table struct {
	Rules []Rule `json:"rules"`
	// Fallback receives anything unmatched. Required: unknown types must go
	// somewhere (I9).
	Fallback string `json:"fallback"`
}

// Router resolves events to route names.
type Router struct {
	// Rules are split into two passes rather than one ordered list, because
	// operation-first precedence is an invariant of the design (I3) and should
	// not depend on someone keeping the ConfigMap in the right order.
	opRules   []Rule // rules that pin an operation
	collRules []Rule // rules that do not
	fallback  string
}

// New builds a Router from a table.
func New(t Table) (*Router, error) {
	if t.Fallback == "" {
		return nil, fmt.Errorf("routing.fallback must be set: unknown types must go somewhere")
	}
	r := &Router{fallback: t.Fallback}
	for i, rule := range t.Rules {
		if rule.Route == "" {
			return nil, fmt.Errorf("routing.rules[%d] has no route", i)
		}
		if rule.Match.Collection != "" {
			if _, err := path.Match(rule.Match.Collection, "probe"); err != nil {
				return nil, fmt.Errorf("routing.rules[%d]: bad collection glob %q: %w",
					i, rule.Match.Collection, err)
			}
		}
		if rule.Match.Operation != "" {
			r.opRules = append(r.opRules, rule)
		} else {
			r.collRules = append(r.collRules, rule)
		}
	}
	return r, nil
}

// Route resolves an event's key to a route name.
//
// Precedence is operation-first, and that is load-bearing (I3):
//
//  1. Rules that pin an operation are evaluated first. `operation == delete`
//     therefore beats the collection map — a deleted post goes to retraction,
//     not content. This is not stylistic: a delete commit carries no record at
//     all, so a create-path handler would have nothing to match against even if
//     it received one. Getting this backwards is the easiest way to break the
//     routing, which is why the test for it is parametrised across the whole
//     table rather than spot-checked.
//
//  2. Then collection rules, in configured order.
//
//  3. Then the fallback. Nothing is ever silently discarded: new lexicons appear
//     on the network continuously, and discovering one should be an observation,
//     not an outage.
func (r *Router) Route(k event.Key) string {
	for _, rule := range r.opRules {
		if matches(rule.Match, k) {
			return rule.Route
		}
	}
	for _, rule := range r.collRules {
		if matches(rule.Match, k) {
			return rule.Route
		}
	}
	return r.fallback
}

func matches(m Match, k event.Key) bool {
	if m.Kind != "" && m.Kind != k.Kind {
		return false
	}
	if m.Operation != "" && m.Operation != k.Operation {
		return false
	}
	if m.Collection != "" {
		if k.Collection == "" {
			return false
		}
		// path.Match treats '.' as an ordinary character and '*' as "any run of
		// non-separator characters", which is the behaviour wanted for NSIDs.
		ok, err := path.Match(m.Collection, k.Collection)
		if err != nil || !ok {
			return false
		}
	}
	return true
}

// Routes lists every route the table can produce, so the runtime builds exactly
// the routes that are reachable and no others.
func (r *Router) Routes() []string {
	seen := map[string]bool{}
	var out []string
	add := func(name string) {
		if name != "" && !seen[name] {
			seen[name] = true
			out = append(out, name)
		}
	}
	for _, rule := range r.opRules {
		add(rule.Route)
	}
	for _, rule := range r.collRules {
		add(rule.Route)
	}
	add(r.fallback)
	return out
}

// Collections lists the concrete (non-glob) collections the table names, for
// Jetstream's server-side filter.
//
// Derived from the table rather than configured beside it, so the filter cannot
// drift out of sync with what we actually know how to route. It is off by
// default: filtering at the source is legitimate load shedding, but discarding
// the type mix removes the problem this service exists to solve, and removes
// any ability to observe what was dropped.
func (r *Router) Collections() []string {
	var out []string
	for _, rule := range r.collRules {
		c := rule.Match.Collection
		if c != "" && !strings.ContainsAny(c, "*?[") {
			out = append(out, c)
		}
	}
	return out
}
